File: app/lib/api/occupied_rooms_api.dart
import '../models/occupied_room.dart';
import 'api_client.dart';

class OccupiedRoomsApi {
  OccupiedRoomsApi(this._client);
  final ApiClient _client;

  /// `GET /api/rooms/occupied?date=&time=` — active markings at a date+time.
  /// [time] is an integer HHMM (e.g. 1430).
  Future<List<OccupiedRoom>> fetchOccupiedRooms({
    required String date,
    required int time,
  }) async {
    final data = await _client.requestJson('/api/rooms/occupied',
        query: {'date': date, 'time': time});
    final raw = data['markings'];
    if (raw is List) {
      return raw
          .whereType<Map>()
          .map((e) => OccupiedRoom.fromJson(Map<String, dynamic>.from(e)))
          .toList();
    }
    return const [];
  }

  Future<OccupiedRoom> createOccupiedRoom({
    required String room,
    required String date,
    required String start,
    required String end,
    String? note,
  }) async {
    final data = await _client.requestJson('/api/rooms/occupied',
        method: 'POST',
        body: {
          'room': room,
          'date': date,
          'start': start,
          'end': end,
          if (note != null && note.isNotEmpty) 'note': note,
        });
    return OccupiedRoom.fromJson(Map<String, dynamic>.from(data['marking']));
  }

  Future<void> removeOccupiedRoom(String id) async {
    await _client.requestJson('/api/rooms/occupied/$id', method: 'DELETE');
  }
}

File: src/utils/occupiedRoomsApi.js
import { apiFetch } from '../auth/AuthContext';

function apiError(res, data) {
    const err = new Error((data && data.error) || `HTTP ${res.status}`);
    err.status = res.status;
    err.code = data && data.error;
    return err;
}

/** Active manual occupancy markings for a calendar date + time (HHMM integer). */
export async function fetchOccupiedRooms({ date, time }) {
    const params = new URLSearchParams({
        date: String(date),
        time: String(time),
    });
    const res = await apiFetch(`/api/rooms/occupied?${params}`);
    const data = await res.json().catch(() => ({}));
    if (!res.ok) throw apiError(res, data);
    return Array.isArray(data.markings) ? data.markings : [];
}

export async function createOccupiedRoom({ room, date, start, end, note }) {
    const payload = { room, date, start, end };
    if (note) payload.note = note;

    const res = await apiFetch('/api/rooms/occupied', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
    });
    const data = await res.json().catch(() => ({}));
    if (!res.ok) throw apiError(res, data);
    return data.marking;
}

export async function removeOccupiedRoom(id) {
    const res = await apiFetch(`/api/rooms/occupied/${encodeURIComponent(id)}`, {
        method: 'DELETE',
    });
    const data = await res.json().catch(() => ({}));
    if (!res.ok) throw apiError(res, data);
    return data.id;
}
